keps/sig-apps/0008-sidecarcontainer.md

```diff
@@ -0,0 +1,116 @@
+* [Implementation History](#implementation-history)
+* [Alternatives](#alternatives-optional)
+
+## Summary
+
+To solve the issue of sidecars not being well supported we can create a new type of container called a sidecar which are handled like normal containers but with the caveat that running sidecar containers will always be terminated when all non-sidecar containers have terminated. This will solve the problem of Jobs not completing/failing when a sidecar container never exits due to it having no awareness of the 'primary' container.
+
+## Motivation
+
+SideCar containers have always been used in some ways but just not formally identified as such, with the introduction of Jobs and Cronjobs it became clear that they're not handled very well. If you have a Job with two containers one of which is actually doing the main processing of the job and the other is just facilitating it, you encounter a problem when the main process finishes; your sidecar container will carry on running so the job will never finish.
+
+The only way around this problem is to manage the sidecar container's lifecycle manually and arrange for it to exit when the main container exits. This is typically achieved by building an ad-hoc signalling mechanism to communicate completion status between containers. Common implementations use a shared scratch volume mounted into all pods, where lifecycle status can be communicated by creating and watching for the presence of files. This pattern has several disadvantages:
+
+* Repetitive lifecycle logic must be rewritten in each instance a sidecar is deployed.
+* Third-party containers typically require a wrapper to add this behaviour, normally provided via an entrypoint wrapper script implemented in the k8s container spec. This adds undesirable overhead and introduces repetition between the k8s and upstream container image specs.
+* The wrapping typically requires the presence of a shell in the container image, so this pattern does not work for minimal containers which ship without a toolchain.
+
+### Goals
+
+Solve issue [25908](https://github.com/kubernetes/kubernetes/issues/25908) so that sidecars can be easily handled without needing modifications to the container itself
+
+### Non-Goals
+
+Changing the way jobs are handled
+
+## Proposal
+
+Create a way to define containers as sidecars, this will be an additional field to the Container Spec: `sidecar: true`.
+e.g:
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: myapp-pod
+  labels:
+    app: myapp
+spec:
+  containers:
+  - name: myapp
+    image: myapp
+    command: ['do something']
+  - name: sidecar
+    image: sidecar-image
+    sidecar: true
+    command: ["do something to help my app"]
+
+```
+These sidecar containers will be handled the same way as a non-sidecar container in almost all respects apart from the fact that when all non-sidecar containers have terminated all running sidecar containers will also be terminated with a reason of `Completed`.
```

Comment:
> running sidecar containers will also be terminated

Based on [@jmillikin-stripe's comment](https://github.com/kubernetes/community/pull/2148#issuecomment-394463716), I think we should clarify a few aspects about the termination behaviour of the pods marked as sidecars.

My thoughts on an approach:

 - For the intended usecase where the mains stop working of their own accord: it would make sense for the sidecars to go through the standard lifecycle, e.g. send a `SIGTERM` when all the non-sidecars have terminated, and provide a grace period for the sidecars to finalise any remaining actions and shut down.  This feels like an (hopefully non-controversial) improvement to tighten up the spec while providing for the sidecars to complete some shutdown activities.
- For non-(Cron)Job-invoked pods which are removed externally (e.g. the pod is deleted), one assumes the main and sidecar containers will continue to be terminated simultaneously with a `SIGTERM`, as now, rather than attempt to stop the mains first followed by the sidecars.
- If all main containers in a pod fail and are, or are in the process of being, restarted, the sidecars should continue to operate within being terminated.

It would also make sense to tighten up the grace period for termination. Using the grace period for shutdown already established and controlled via the `terminationGracePeriodSeconds` value in the Pod spec would make sense to me. In non-(Cron)Job tasks, the sidecars would have the full grace period to terminate (as shutdown only commences when the mains have terminated). For pods which are removed externally, the above suggests mains and sidecars terminate simultaneously, so the grace period again applies to both equally.

This approach does not provide the ability to implement a deferred shutdown behaviour in all circumstances, which [@fejta's comment](https://github.com/kubernetes/community/pull/2148#issuecomment-394508321) mentioned. That feels like it is out-of-scope for this proposal.

Reply:
If we guarantee that the pod will get a `SIGTERM` or something like it when the primary containers finish, and we have `terminationGracePeriodSeconds` to complete any outstanding work that will meet our needs

Reply:
Yeah I agree, I think they should be handled the same way as a normal containers as much as possible, it makes life a lot simpler. I'll add a bit clarifying how their termination is handled.

Reply:
> If all main containers in a pod fail and are, or are in the process of being, restarted, the sidecars should continue to operate within being terminated.
That's actually an interesting point, I thought this wouldn't be a problem as CrashLoopBackoff counts as a Waiting state, but before that it will briefly have state Terminated. So you'd need some kind of way of knowing that the other containers have terminated permanently. I guess you'd just use the pod restart policy to work it out.